Ddvo contrib5: improvements to est_client.c/verify_cacert_resp() and Makefiles

DDvO opened the pull request:
As requested, removed inefficiencies and memory leak from verify_cacert_resp()
I also simplified its deallocation code, curing wrong deallocation order
- as written in http://linux.die.net/man/3/x509_store_ctx_init
The certificates and CRLs in a context are used internally and should not be freed up until after the associated X509_STORE_CTX is freed.

While compiling the example estclient etc, I found again that their Makefiles do not have a reference to libest, such that they are not re-built by a local "make" when libest was updated. I added the missing references, also for the unit tests, and in the course of doing so, I removed some duplicate compiler and linker arguments.

rpb5bnc commented:
One issue at a time.  There are at least two reasons for this,

1. Combining multiple issues together makes it harder to review each one on its own merit and turns the merging effort into a much larger project for the code owner.

2. Let's say a pull request is made that contains code changes for four distinct issues.  It turns out that only one is to be accepted.  It now becomes the code owners responsibility to sift through the diffs trying to figure out which changes go with the one issue that's being accepted.

It could be argued that a request with only a few small independent changes could be handled, and that's true, it could.  But it's been found that it's much better to move deliberately and methodically in maintaining code, especially code that is security related.  Accepting a pull request with even a few small independent issue resolutions also begins to get into a slippery slope situation.  The definition of small and manageable becomes subjective.  If the code owner accepts one with a few small independent changes, and then rejects someone else's request containing multiple independent issues problems can arise.

So, one change at a time.

All this being said, a few comments about these issues.

It's stated that the example applications, estclient, estserver, and estproxy, are not getting recompiled when a change is made to the library.  What was the change that was made to libest that did not cause a rebuild of the example applications?

DDvO commented:
Accordingly, I
Accordingly, I've split the branches ddvo_contrib5 into two:
* ddvo_contrib5b (on est_client.c verify_cacert_resp())
* ddvo_contrib5c (on improvements of the example/Makefiles)
(see https://github.com/DDvO/libest/branches for details),
such that I'm now prepared to submit independent pull requests for each.

Or would you like me to break down the first one further (at which line, e.g., performance improvements vs. cleanup of deallocation etc.)?

Concerning your question regarding rebuild:
The problem was that when the libest binary in ../../src/est/.libs/ changed
and "make" is done for any of the example apps or unit tests, they are not re-linked.
The proposed improvement (now moved into ddvo_contrib5c) cures this problem.
This new independent branch meanwhile cures an even more subtle issue:
When any of the sources of the libest core changes and "make" is done for any of the example apps or unit tests, the libest binary is not re-built (and consequently the app/test not re-linked).

rpb5bnc commented:
And why do we believe that just because libest (code under src/est/) was rebuilt, that the example applications (code under example/) need to be rebuilt?  I have to ask once again, what changed in libest that makes you believe the example apps have to be rebuilt?  What file?  Because it makes a difference.

libest is a standard automake based project.  You run make from the top level directory.  Running make from one of the example/ directories will not cause the code under src/est/ to be rebuilt.  This is all working as intended as a standard hierarchical make based project.

DDvO commented:
Whatever change has been done in libest (in any of the core sources; in the above case: the changes to verify_cacert_resp() in src/est/est_client.c), all code depending on it (in the above case: example/client/estclient.c, which I used for testing my changes), needs to be re-built. I wonder why you question that.

Though libEST is an automake project, its setup w.r.t. the dependencies of the various libraries is simply wrong: It acts as if src/est, example/* and test/UT/* were independent, which is not the case, since all other directories depend on /src/est/.

Suppose there is any change to any source file of the core in src/est/. Even when invoking "make" from the top level directory, only src/est/ is re-built. Instead, all dependent code (i.e., the examples and the unit tests) should be re-linked when the core library changed (regardless where "make" was invoked to update the binaries in the current directory).

The proposed update of the Makefiles partly cures this general problem, namely for the case "make" is invoked within any of the dependent directories.

rpb5bnc commented:
No, there is absolutely nothing wrong with this setup.  libest is a standard linux shared library.  The purpose of a library is to abstract away the details, the implementation, of a function, in this case, EST, that is RFC 7030.  For libest, that implementation resides in src/est/.  Every library must have a defined interface to it.  In the case of libest, that interface is defined in est.h.  Code that interacts with libest is dependent upon the interface to libest, NOT the internal logic of the library.  SO, when the internal logic is modified, for example, a claimed memory leak in a function, if that change does NOT modify the interface, in this case est.h, that change can be made without requiring the rest of the system to be rebuilt.  This is what a library is for, AND, this is why I repeatedly asked what file was changed in libest that makes one believe that the rest of the system needs to be rebuilt.  If est.h was changed, then yes, modules that make use of libest need to be rebuilt.  This is exactly how the logic in examples/ and test/ work, so, this repository that implements a linux shared library and additional logic to test and exemplify how one could use the library is all set up to build correctly.

So, one can make a change to say est_client.c and then run make, from the top level.  The library, and only the library, will be rebuilt.  Then, when an application is invoked, such as the four found under example/ which link in libest as a shared library, they will be dynamically linked with the new version.  Because they interact with libest through its defined interface, est.h, and est.h has not changed, the application will still be able to correctly utilize libest and will pick up the changes that were made to it.

Again, this is all standard linux shared library concepts.  libest implements the EST protocol (RFC 7030), and this resides in /src/etc/.  This is not the core of EST, this IS EST.  Any other logic in this repository is either logic to test the library or example code that shows how one could interact with the EST library to make it perform the EST protocol.

DDvO commented:
Ah, now I got your point!
Sure, when assuming dynamic linking, you are (generally) right.

But how can one assume that? In my case (at least when building under a default Cygwin environment), I happen do get static linking, and my above remarks (implicitly, sorry!) assumed this.
How to solve the problem then in case one uses static linking?

I just confirmed the dependency on src/est/est.h for the examples,
but this does NOT hold for test/UT/Makefile (which is not generated by automake).
Here, we at least need to extend
`runtest: $(OBJ)`
to
`runtest: $(OBJ) ../../src/est/est.h`
Also in this case the problem of getting the dependencies right for static linking remains.

(As contained in my pull request/branch ddvo_contrib5c), I also propose the following change:
`LDFLAGS += -lcunit -ldl -lpthread -lssl -lcrypto -lest -lcurl`
to
`LDFLAGS += -lcunit -ldl -lpthread -lest -lssl -lcrypto -lcurl`
since otherwise I get linking errors with gcc.

Moreover, I found that in each example/*/Makefile.am,
`AM_CFLAGS = -I../.. -I$(srcdir)/../../src/est -I$(SSL_CFLAGS) -g`
can be simplified to
`AM_CFLAGS = -I../.. -I$(srcdir)/../../src/est`

and
`estclient_LDFLAGS = -L../../src/est/.libs $(DL) $(PTHREAD) -lest -lssl -lcrypto`
can be simplified to
`estclient_LDFLAGS = -L../../src/est/.libs -lest`

since otherwise the respective compiler/linker arguments are given twice.
